crates/ruff_server/src/server/api/notifications/did_change_watched_files.rs

```diff
@@ -15,13 +16,21 @@ impl super::SyncNotificationHandler for DidChangeWatchedFiles {
     fn run(
         session: &mut Session,
         _notifier: Notifier,
+        requester: &mut Requester,
         params: types::DidChangeWatchedFilesParams,
     ) -> Result<()> {
-        for change in params.changes {
+        for change in &params.changes {
             session
                 .reload_settings(&change.uri)
                 .with_failure_code(lsp_server::ErrorCode::InternalError)?;
         }
+
+        if session.resolved_client_capabilities().workspace_refresh && !params.changes.is_empty() {
+            requester
+                .request::<types::request::WorkspaceDiagnosticRefresh>((), |()| Task::nothing())
+                .with_failure_code(lsp_server::ErrorCode::InternalError)?;
+        }
```

Comment:
I think we should show a warning to users if the `workspace_refresh` capapbiltiy is not supported, telling them that they might see stale diagnostics and should restart the server. Or is there another way for us to push diagnostics (or clear them)?

Related to my question on the show error message PR. Why don't we show an error message here if senting the refresh request failed?

Reply:
Neovim doesn't support this capability and I don't know of any server which would send a message to the client so I'm not sure if it's ok to do so. Rust analyzer does refresh the workspace when any of the `Cargo.toml` file changes but that's probably something else.

Reply:
What do you mean by refreshing the workspace? Is this another LSP command that we could use as fallback?

Reply:
@MichaReiser At the moment we can't even check whether or not the capability is supported due to a [bug in `lsp-types`](https://github.com/gluon-lang/lsp-types/pull/281). I recommend we do this in a follow-up issue.

> Why don't we show an error message here if senting the refresh request failed?

The callback only runs if the request succeeds. Any erroneous response is [handled internally](https://github.com/astral-sh/ruff/blob/main/crates/ruff_server/src/server/client.rs#L165-L168).

Reply:
> What do you mean by refreshing the workspace? Is this another LSP command that we could use as fallback?

I'd need to look into it but whenever I change any `Cargo.toml` files, the server starts indexing the project again. Maybe the server is watching over those files and starting to refresh all the relevant things like semantic tokens, inlay hints and possibly diagnostics.